### PR TITLE
<Button/> - Tooltip doesn't work on a disabled button

### PR DESCRIPTION
**Pull request by Faradey27**
### What changed

<Button/> - Tooltip doesn't work on a disabled button

### Why it changed

https://github.com/wix/wix-style-react/issues/832

---

- [x] Change is tested
- [x] Change is documented
- [ ] Build is green
- [ ] Change of UX is approved by [wuwa](https://github.com/wuwa) or [milkyfruit](https://github.com/milkyfruit)

### CODE REVIEW

**Comment by Faradey27** [comment]
@lbelinsk updated with e2e and unit tests

File: src/Backoffice/Button/Button.driver.js [hunk 1]
```diff
@@ -5,6 +5,10 @@ import ReactTestUtils from 'react-dom/test-utils';
 export default ({element, wrapper, component}) => ({
   exists: () => !!element,
   click: () => ReactTestUtils.Simulate.click(element),
+  focusableOnFocus: () => ReactTestUtils.Simulate.focus(element),
```
Comment (shlomitc) [inline src/Backoffice/Button/Button.driver.js]: maybe just `focus`?

File: src/Backoffice/Button/Button.driver.js [hunk 2]
```diff
@@ -5,6 +5,10 @@ import ReactTestUtils from 'react-dom/test-utils';
...
+  focusableOnBlur: () => ReactTestUtils.Simulate.blur(element),
```
Comment (shlomitc) [inline src/Backoffice/Button/Button.driver.js]: `blur`?

File: src/Backoffice/Button/Button.driver.js [hunk 3]
```diff
@@ -5,6 +5,10 @@ import ReactTestUtils from 'react-dom/test-utils';
...
+  onMouseEnter: () => ReactTestUtils.Simulate.mouseEnter(element),
```
Comment (shlomitc) [inline src/Backoffice/Button/Button.driver.js]: `mouseEnter`

File: src/Backoffice/Button/Button.driver.js [hunk 4]
```diff
@@ -5,6 +5,10 @@ import ReactTestUtils from 'react-dom/test-utils';
...
+  onMouseLeave: () => ReactTestUtils.Simulate.mouseLeave(element),
```
Comment (shlomitc) [inline src/Backoffice/Button/Button.driver.js]: `mouseLeave`

File: src/Backoffice/Button/Button.js [hunk 5]
```diff
@@ -63,13 +63,13 @@ class Button extends WixComponent {
...
         <button
-          onClick={onClick}
+          onClick={disabled ? null : onClick}
           disabled={disabled}
           type={type}
           onMouseEnter={onMouseEnter}
           onMouseLeave={onMouseLeave}
-          onFocus={this.props.focusableOnFocus} // For some reason eslint react/prop-types rule doesn't work here ?!#$
-          onBlur={this.props.focusableOnBlur}
+          onFocus={disabled ? null : this.props.focusableOnFocus} // For some reason eslint react/prop-types rule doesn't work here ?!#$
```
Comment (shlomitc) [inline src/Backoffice/Button/Button.js]: is this being currently called even when the button is disabled?

File: src/Tooltip/Tooltip.e2e.js [hunk 6]
```diff
@@ -11,6 +11,16 @@ describe('Tooltip', () => {
     browser.get(storyUrl);
   });
 
+  eyes.it('should show tooltip when button disabled', async () => {
+    browser.executeScript('window.setComponentState({disabledButtonWithTooltip: true})');
```
Comment (shlomitc) [inline src/Tooltip/Tooltip.e2e.js]: @argshook - have a better suggestion for that?

File: stories/Tooltip/Composite/TooltipForEyesOnly.js [hunk 7]
```diff
@@ -15,28 +15,58 @@ export default class TooltipForEyesOnly extends Component {
     super();
     this.state = {
       content: SHORT_CONTENT,
+      disabledButtonWithTooltip: false,
```
Comment (shlomitc) [inline stories/Tooltip/Composite/TooltipForEyesOnly.js]: this is more of a button issue than tooltip, so maybe it should be actually a button events issue. I think that testing it in a unit-test is enough

File: src/Backoffice/Button/Button.driver.js (hunk 1 above)
Comment (Faradey27) [reply src/Backoffice/Button/Button.driver.js]: in component api it called focusableOnFocus, so i decided to left same)
@shlomitc

File: src/Backoffice/Button/Button.driver.js (hunk 2 above)
Comment (Faradey27) [reply src/Backoffice/Button/Button.driver.js]: in component api it called focusableOnBlur, so i decided to left same)
@shlomitc

File: src/Backoffice/Button/Button.driver.js (hunk 3 above)
Comment (Faradey27) [reply src/Backoffice/Button/Button.driver.js]: in component api it called onMouseEnter, so i decided to left same)
@shlomitc

File: src/Backoffice/Button/Button.driver.js (hunk 4 above)
Comment (Faradey27) [reply src/Backoffice/Button/Button.driver.js]: in component api it called mouseLeave, so i decided to left same)
@shlomitc

File: src/Backoffice/Button/Button.driver.js (hunk 1 above)
Comment (shlomitc) [reply src/Backoffice/Button/Button.driver.js]: the driver should be an abstraction for the functionality, meaning the consumer will want to trigger a focus event on the component, so it will make it more declarative to call it `focus`

File: stories/Tooltip/Composite/TooltipForEyesOnly.js (hunk 7 above)
Comment (Faradey27) [reply stories/Tooltip/Composite/TooltipForEyesOnly.js]: @shlomitc agree about button related

but not agree about e2e test, because main change was done in css, by removing pointer-events: none, and such case could not be handled by jsdom(currently)

File: src/Backoffice/Button/Button.js (hunk 5 above)
Comment (Faradey27) [reply src/Backoffice/Button/Button.js]: no, it was not

**Comment by Faradey27** [comment]
@shlomitc could you approve eyes and merge?

File: src/Backoffice/Button/Button.js (hunk 5 above)
Comment (shlomitc) [reply src/Backoffice/Button/Button.js]: so why do you need to add the check for the disabled state? doesn't the native button disables it?

File: src/Backoffice/Button/Button.protractor.driver.js [hunk 8]
```diff
@@ -10,6 +11,10 @@ const buttonDriverFactory = element => {
 
   const publicDriver = {
     click: () => element.click(),
+    hover: element => mouseEnter(element),
+    tooltipElement: () => $('[data-hook="tooltip-e2e-wrapper"]'),
```
Comment (shlomitc) [inline src/Backoffice/Button/Button.protractor.driver.js]: please don't! this is not part of the Button component, I really don't want to have such functionality in the driver.
The driver should be a simple abstraction to let consumer interact with the component in tests

same for the next functions related to the tooltip

File: stories/Button/standard.story.js [hunk 9]
```diff
@@ -24,5 +28,11 @@ export default {
     onMouseLeave: () => 'Mouse Leave!',
     prefixIcon: icons,
     suffixIcon: icons
-  }
+  },
+
+  examples: (
+    <CodeExample title="Disabled button with tooltip" code={ExampleDisabledButtonWithTooltipRaw}>
+      <ExampleDisabledButtonWithTooltip/>
```
Comment (shlomitc) [inline stories/Button/standard.story.js]: @argshook - What do you think about this kind of visual edge cases E2E tests?

The fix is very small and is CSS solution so It does require a browser test.

In my opinion, testing the hover functionality (even without a tooltip) will be enough but anyway, this means we pollute our API page with test-only examples.

I'll be happy to hear another voice in this issue

**Comment by Faradey27** [comment]
@shlomitc i removed e2e test

File: src/Backoffice/Button/Button.js (hunk 5 above)
Comment (shlomitc) [reply src/Backoffice/Button/Button.js]: I mean that when the native `button` gets `disabled` prop, it automatically stops triggering the focus and blur events.
for example (see the console logs):
http://jsbin.com/toqapigoga/1/edit?html,js,console,output

File: stories/Button/standard.story.js (hunk 9 above)
Comment (argshook) [reply stories/Button/standard.story.js]: i suppose such examples shouldn't be visible to end user and so we need to hide them. currently we don't have mechanism for that specificly

some thoughts
1. create separate story page for e2e only examples and somehow hide it (skimmed through storybook api, didn't see a way to hide story from sidebar, but im pretty sure it's possible, need to dig more)
2. improve wix-storybook-utils to support something like `examplesE2e` or similar to live nearby `examples`. These would be displayed only in e2e (and so current `withExamples` in url wouldn't be needed anymore)

File: src/Backoffice/Button/Button.js (hunk 5 above)
Comment (Faradey27) [reply src/Backoffice/Button/Button.js]: yea, you are completely right

**Comment by Faradey27** [comment]
@shlomitc updated
